src/modules/imaging/location.py

```diff
@@ -0,0 +1,178 @@
+from dataclasses import dataclass
+import math
+from src.modules.imaging.mavlink import MAVLinkDelegate
+
+
+@dataclass
+class LatLng:
+    """
+    Latitude and longitude in WGS84 coordinates.
+    """
+
+    lat: float
+    lng: float
+
+
+@dataclass
+class Heading:
+    """
+    Heading, as defined in an aeronautical sense, with respect to
+    north. Measured in degrees.
+    """
+
+    heading: float
+
+
+@dataclass
+class Rotation:
+    """
+    Orientation of the drone using euler angles in degrees.
+    """
+
+    pitch: float
+    roll: float
+    yaw: float
+
+
+class LocationProvider:
+    """
+    Provides the drone's current location and orientation as read from
+    telemetry or a debug-source.
+    """
+
+    def location(self) -> LatLng:
+        """
+        Get the latest lat/lng location of the drone.
+        """
+        raise NotImplementedError()
+
+    def heading(self) -> Heading:
+        """
+        Get the latest heading of the drone (with respect to north.)
+        """
+        raise NotImplementedError()
+
+    def altitude(self) -> float:
+        """
+        Get the latest altitude of the drone in meters.
+        """
+        raise NotImplementedError()
+
+    def orientation(self) -> Rotation:
+        """
+        Get the latest 3D orientation of the drone in it's inertial frame.
+        """
+        raise NotImplementedError()
+
+
+class DebugLocationProvider:
+    """
+    Will return a series of given locations and orientations.
+    """
+
+    def __init__(self) -> None:
+        self._current_location = LatLng(0.0, 0.0)
+        self._current_heading = Heading(0.0)
+        self._current_altitude = 0.0
+        self._current_orientation = Rotation(0.0, 0.0, 0.0)
+
+    def location(self) -> LatLng:
+        return self._current_location
+
+    def heading(self) -> Heading:
+        return self._current_heading
+
+    def altitude(self) -> float:
+        return self._current_altitude
+
+    def orientation(self) -> Rotation:
+        return self._current_orientation
+
+    def set_location(self, new_location):
+        self._current_location = new_location
+
+    def set_heading(self, new_heading):
+        self._current_heading = new_heading
+
+    def set_altitude(self, new_altitude):
+        self._current_altitude = new_altitude
+
+    def set_orientation(self, new_orientation):
+        self._current_orientation = new_orientation
+
+    def debug_change_location(self, **kwargs):
+        """
+        Change the location reported by this DebugLocationProvider.
+
+        Keyword Arguments:
+        lat -- Latitude
+        lng -- Longitude
+        heading -- Heading direction
+        altitude -- Altitude
+        pitch -- Pitch angle
+        roll -- Roll angle
+        yaw -- Yaw angle
+        """
```

Comment:
This API is really nice! There is an issue if we don't pass a full combination of some arguments:

```python
debug_log = DebugLocationProvider()
debug_loc.debug_change_location(pitch=12)
debug_loc.orientation().pitch  # => 0?
```

It would be nice if instead of requiring a "full set" of fields to update the location, or orientation, we instead automatically merged the updates. Not sure how much sense that makes, so here's an example:

```python
debug_log = DebugLocationProvider()
debug_loc.orientation().pitch  # => 0
debug_loc.orientation().roll  # => 0
debug_loc.orientation().yaw  # => 0
debug_loc.debug_change_location(pitch=12)
debug_loc.orientation().pitch  # => 12
debug_loc.orientation().roll  # => 0
debug_loc.orientation().yaw  # => 0
```

In terms of implementation:

```python
if any(k in kwargs for k in ['pitch', 'roll', 'yaw']):
  pitch = kwargs.get('pitch', default=self._current_orientation.pitch)
  roll = ...
  yaw = ...
  self._current_orientation = Rotation(pitch, roll, yaw)
```

Notice the `any` instead of `all` and that we use a `default` if the user didn't pass a pitch and, eg. only a yaw.